### PR TITLE
Ridhwana/Load articles dynamically for shecoded

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
For context the SheCoded page will be created from `/internal/pages`, and that is where the html and styles will get uploaded. 

Once uploaded, the page is created then we can use javascript to get the articles for the campaign and render it dynamically in the div.

The file is created is specifically used for SheCoded, we can try to make it more general for campaigns in the future once we know more (classnames, params).

## Related Tickets & Documents
https://github.com/thepracticaldev/dev.to/issues/6260

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
Still busy polishing the design but the articles are dynamic in the screenshot below.

## Added tests?

- [ ] yes
- [x] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+// This file is specifically used for SheCoded,
+// We can try to make it more general for campaigns in the future once we know more (classnames, params).
+
+function renderArticles(articles, articleElement ) {
```
Comment (rhymes) [inline app/assets/javascripts/initializers/initializeSheCodedArticles.js]: small typo, the space

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  articleDiv.innerHTML = articleHTML.join(" ");
+}
+
+function renderNoArticles(articleElement) {
+  let articleDiv = articleElement;
+  articleDiv.innerHTML = "<div>There are no articles to show.</div>";
+}
+
+function getArticles(articleElement) {
```
Comment (rhymes) [inline app/assets/javascripts/initializers/initializeSheCodedArticles.js]: I think [fetch()](https://javascript.info/fetch) would result in a lot less code here. We already use it around the JS code if you need examples.

**Comment by Ridhwana** [comment]
> How can I test this? Can't find the page with `js-she-coded-page__some-stories-sub-container`

`For context the SheCoded page will be created from /internal/pages, and that is where the html and styles will get uploaded.`

@rhymes The only way to test it is to create an internal page and pop some html with a high level div name `js-she-coded-page__some-stories-sub-container`.😞This is how it will work in prod.

**Comment by benhalpern** [comment]
I think directly inlining the JavaScript as a `<script>` tag in the html of database-backed page itself would be my preferred approach. I'm cool with this if you think it's better for the problem at hand, as the most important thing is we just make it happen.

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js (hunk 1 above)
Comment (vaidehijoshi) [reply app/assets/javascripts/initializers/initializeSheCodedArticles.js]: i feel like eslint should have caught this, right? 👀

Comment (rhymes) [reply app/assets/javascripts/initializers/initializeSheCodedArticles.js]: Don't know, maybe!  We do have 2 sets of "eslint config files", one for these "legacy" JS files and one for the component oriented Preact files. Maybe their configurations are not aligned

cc @nickytonline for when he gets back from the holidays

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js (hunk 2 above)
Comment (vaidehijoshi) [reply app/assets/javascripts/initializers/initializeSheCodedArticles.js]: +1, we should definitely be using fetch here

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  let xmlhttp;
+  if (window.XMLHttpRequest) {
+    xmlhttp = new XMLHttpRequest();
+  } else {
+    xmlhttp = new ActiveXObject("Microsoft.XMLHTTP");
+  }
+
+  xmlhttp.onreadystatechange = function() {
+    if (xmlhttp.readyState === XMLHttpRequest.DONE) {
```
Comment (vaidehijoshi) [inline app/assets/javascripts/initializers/initializeSheCodedArticles.js]: using fetch here will also let you use the promise api here

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            <img src="${article.user.profile_image}" alt="${article.user.name}'s profile picture"/>
+          </div>
+          <div class="she-coded-page__some-story-info-name">${article.user.name}</div>
+        </div>
+        <div class="she-coded-page__some-story-info-date">&nbsp; &#183; &nbsp; ${article.readable_publish_date}</div>
+      </div>
+    </div></a>`
+  });
+
+  articleDiv.innerHTML = articleHTML.join(" ");
```
Comment (vaidehijoshi) [inline app/assets/javascripts/initializers/initializeSheCodedArticles.js]: I'm pretty sure react expects `dangerouslySetInnerHTML` here. Some relevant [docs](https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml) here.

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+function renderNoArticles(articleElement) {
+  let articleDiv = articleElement;
+  articleDiv.innerHTML = "<div>There are no articles to show.</div>";
```
Comment (vaidehijoshi) [inline app/assets/javascripts/initializers/initializeSheCodedArticles.js]: Same comment re: using `innerHTML` here.

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  let url = "/api/articles";
+  let params = "tag=shecoded&per_page=12";
+
+  xmlhttp.open("GET", `${url}?${params}`, true);
+  xmlhttp.send();
+}
+
+function initializeSheCodedArticles() {
+  let articleElement = document.getElementById('js-she-coded-page__some-stories-sub-container');
+  if(articleElement) {
```
Comment (vaidehijoshi) [inline app/assets/javascripts/initializers/initializeSheCodedArticles.js]:
```suggestion
  if (articleElement) { getArticles(articleElement); }
```

**Comment by Ridhwana** [comment]
> I think directly inlining the JavaScript as a `<script>` tag in the html of database-backed page itself would be my preferred approach. I'm cool with this if you think it's better for the problem at hand, as the most important thing is we just make it happen.

@benhalpern I did not think about this 🤦‍♀ , it makes a lot more sense to have the JS close to the HTML.

**Comment by vaidehijoshi** [comment]
Just left a few comments about using the the `fetch` API and React's replacement for `innerHTML`!

File: app/assets/javascripts/initializers/initializeSheCodedArticles.js (hunk 4 above)
Comment (rhymes) [reply app/assets/javascripts/initializers/initializeSheCodedArticles.js]: this is not a component though, it's "free form", old school javascript :)

Comment (Ridhwana) [reply app/assets/javascripts/initializers/initializeSheCodedArticles.js]: @vaidehijoshi we're not using preact here, its just a plain js initializer

Comment (vaidehijoshi) [reply app/assets/javascripts/initializers/initializeSheCodedArticles.js]: Ahhhh, whoops! You're right. Please ignore this, entirely.

**Comment by Ridhwana** [comment]
Hi everyone! Thanks for taking the time to review this PR. However, the better approach here is to directly inline the JavaScript as a <script> tag with the html and css, instead of having a random script in the codebase! I'll still use some of the suggestions here though, thanks!

**Pull request closed without merging**